[localize] Auto generate message ids

aomarks opened the pull request:
The `options.id` parameter is now optional, and when omitted, an ID is automatically generated by hashing the string contents of the template.

Example:
  - Template: ``` html`Hello <b>${who}</b>!` ```
  - Hashed strings: `["Hello <b>", "</b>!"]`
  - Output: `ah82ccc38d4d46eaa9`

 The ID is constructed as:

   - `[0]`    Version indicator this ID generation scheme (`a`). Distinguishes these IDs from different schemes we might use in the future.
   - `[1]`    Kind of template: [`h`]tml or [`s`]tring. Distinguishes e.g. ``` `<b>foo</b>` ``` from ``` html`<b>foo</b>` ```.
   - `[2,17]` 64-bit FNV-1a hash hex digest of the template strings, where each string is UTF-8 encoded and delineated by an ASCII "record separator" character.

 I chose FNV-1a because:

   1. It's pretty fast. Much faster than SHA-1 or MD5, or probably any cryptographically secure one. I compared against a whole bunch of other non-cryptographic ones I found, and while this wasn't the *very* fastest, it seemed to be the best choice on balance (can go into more detail if interested).
   2. It's pretty small (0.41 KiB minified + brotli).
   3. We don't require cryptographic security, and 64 bits should give sufficient collision resistance for any one application. I did a non-scientific collision tests on ~5 million strings (every line in a portion of the Wikipedia XML dump, and every line in Shakespeare), and found none. Worst case, we will always detect collisions during analysis.
   4. We can't use Web Crypto API (e.g. SHA-1), because it's asynchronous.
   5. It's a [well known](https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function) non-cryptographic hash with implementations in many languages (e.g. [Go stdlib](https://golang.org/pkg/hash/fnv/)). I tested the ~5 million strings against the stdlib Go implementation to get some confidence it's correct, and they all matched.
   6. There was an [existing JavaScript implementation](https://github.com/tjwebb/fnv-plus) that doesn't require `BigInt`, for IE11 compatibility, and has ~7K monthly NPM downloads.
  7. We can switch to https://github.com/sindresorhus/fnv1a/blob/master/index.js#L47, which is really tiny, when we eventually drop IE11 support (needs `BigInt`).

## Benchmarks

These benchmarks compare runtime performance between:

- auto: ``` msg(html`foo`) ```
- baked: ``` msg(html`foo`, {id: 'someid'}) ```
- transformed: ``` html`foo` ```

(Note that this "baked" style is something that the user would currently need to do manually, but in the future we could automatically generate it for them as a third output style).

First render of 100 templates:

![image](https://user-images.githubusercontent.com/48894/99891287-1270fb00-2c1d-11eb-94f7-253ceb26cae1.png)

100 *subsequent* renders of 100 templates:

![image](https://user-images.githubusercontent.com/48894/99891290-169d1880-2c1d-11eb-9f5b-26d025d607eb.png)

Note this shows that, since we cache each template's hash after first render, the slowdown from auto ids is mostly gone after the first render (see baked vs auto in the second table).

Fixes https://github.com/Polymer/lit-html/issues/1437

justinfagnani commented:
I made a template result has function for using with SSR here: https://github.com/Polymer/lit-html/blob/lit-next/packages/lit-html/src/hydrate.ts#L392

It'd be nice to analyze that for perf and collisions like you did here, and use the same hash if possible.

justinfagnani commented on packages/localize/src_client/lit-localize.ts:
```diff
@@ -346,4 +348,24 @@ export function _msg(
+    // Note that by using HASH_DELIMITER as the template parameter here, we can
+    // skip splitting and re-joining when we perform the hash.
+    const params = Array(numParams).fill(HASH_DELIMITER);
+    template = template(...params);
```
Another option here, that would allow expressions in templates, would be to evaluate the template function with the provided args, then pull the strings from the template result.

aomarks replied:
> fascinating :) - it's safe to pass this here because expressions are required to be simple identifiers that only reference the parameters. Might be nice to comment that for future maintainers

Done.

> Another option here, that would allow expressions in templates, would be to evaluate the template function with the provided args, then pull the strings from the template result.

Yeah, that occurred to me while working on this too! I like idea, but there are a few details to account for I thought of:

1. Calls like ``` msg(`Hello ${who}!`) ``` won't work, because we'd just receive a string. I think we'd need to vend a tag like ``` str`Hello ${who}!` ``` for this case so that we can capture the strings and values.

2. The order of placeholders can change during translation, so we'd need to analyze their new locations and include that re-ordering logic into the functions we generate.

3. A side-effect advantage of identifiers is that I think they'll tend to be easier for translators to understand, since they'll be more self-describing. Translators see the expressions (but can't change them), to help them understand context. I think churn in the values is also less likely with identifiers.

aomarks reviewed (commented):
> I made a template result has function for using with SSR here: https://github.com/Polymer/lit-html/blob/lit-next/packages/lit-html/src/hydrate.ts#L392
> 
> It'd be nice to analyze that for perf and collisions like you did here, and use the same hash if possible.

Talked offline. The FNV-1a hash used here has much better collision resistance, and is faster. It's bigger, though. I'm still not super confident about the choice of hash function here, I'd like something smaller. Will probably think about this a bit more before releasing the next version.

aomarks commented:
I added a small change: switched from `0` to `a` as the ID scheme prefix, because that will let minifiers use an identifier instead of a quoted string in generated translation objects.
